pkg/dao/rpm.go

```diff
@@ -170,3 +173,166 @@ func (r rpmDaoImpl) Search(orgID string, request api.SearchRpmRequest, limit int
+	return result, found
+}
+
+func (r rpmDaoImpl) fetchRepo(uuid string) (models.Repository, error) {
+	found := models.Repository{}
+	result := r.db.
+		Where("UUID = ?", uuid).
+		First(&found)
+
+	if result.Error != nil {
+		return found, result.Error
+	}
+	return found, nil
+}
+
+func (r rpmDaoImpl) InsertForRepository(repoUuid string, pkgs []yum.Package) (int64, error) {
+	var rowsAffected int64
+
+	repo, err := r.fetchRepo(repoUuid)
+	if err != nil {
+		return rowsAffected, err
+	}
+	checksums := make([]string, len(pkgs))
+	for i := 0; i < len(pkgs); i++ {
+		checksums[i] = pkgs[i].Checksum.Value
+	}
+
+	result, existingChecksums := r.existingChecksums(checksums)
+	if result.Error != nil {
+		return rowsAffected, result.Error
+	}
+
+	dbPkgs := FilteredConvert(pkgs, existingChecksums)
+
+	rowsAffected, error := r.PagedRpmInsert(&dbPkgs)
+	if error != nil {
+		return rowsAffected, result.Error
+	}
+
+	var rpmUuids []string
+	//Now fetch the uuids of all the rpms we want associated to the repository
+	result = r.db.Where("checksum in ?", checksums).Model(&models.Rpm{}).Pluck("uuid", &rpmUuids)
+	if result.Error != nil {
+		return rowsAffected, result.Error
+	}
+
+	//Delete RepositoryRpm entries we don't need
+	error = r.deleteUnneeded(repo, rpmUuids)
+	if error != nil {
+		return rowsAffected, result.Error
+	}
+
+	//Add the RepositoryRpm entries we do need
+	associations := prepRepositoryRpms(repo, rpmUuids)
+	result = r.db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "repository_uuid"}, {Name: "rpm_uuid"}},
+		DoNothing: true}).Create(&associations)
+
+	return rowsAffected, result.Error
+}
+
+func prepRepositoryRpms(repo models.Repository, rpm_uuids []string) []models.RepositoryRpm {
+	repoRpms := make([]models.RepositoryRpm, len(rpm_uuids))
+	for i := 0; i < len(rpm_uuids); i++ {
+		repoRpms[i].RepositoryUUID = repo.UUID
+		repoRpms[i].RpmUUID = rpm_uuids[i]
+	}
+	return repoRpms
+}
+
+// difference returns the difference between arrays a and b   (a - b)
+func difference(a, b []string) []string {
+	mb := make(map[string]struct{}, len(b))
+	for _, x := range b {
+		mb[x] = struct{}{}
+	}
+	var diff []string
+	for _, x := range a {
+		if _, found := mb[x]; !found {
+			diff = append(diff, x)
+		}
+	}
+	return diff
+}
+
+func (r rpmDaoImpl) deleteUnneeded(repo models.Repository, rpm_uuids []string) error {
+	//First get uuids that are there:
+	var existing_rpm_uuids []string
+
+	r.db.Model(&models.RepositoryRpm{}).Where("repository_uuid = ?", repo.UUID).Pluck("rpm_uuid", &existing_rpm_uuids)
+
+	rpmsToDelete := difference(existing_rpm_uuids, rpm_uuids)
+
+	//result := db.DB.Where("repositories_rpms.repository_uuid = ? and repositories_rpms.rpm_uuid not in ?", repo.UUID, *rpm_uuids).Delete(&models.RepositoryRpm{})
+	result := r.db.Where("repositories_rpms.repository_uuid = ? and repositories_rpms.rpm_uuid in ?", repo.UUID, rpmsToDelete).Delete(&models.RepositoryRpm{})
```

Comment:
the reason it was "not in UUID_ARRAY"   is that it was passing the list of UUIDs that we want associated to the repo into the function.  So this is meant to delete the ones that aren't in that array.

For example, say a repo already has  [a,b,c]  and then we introspect again, and the list is now [b,c,d].  We'd want to delete anything that isn't in [b,c,d].

Reply:
ignore me, i need to get my eyes checked